### PR TITLE
Use syntax highlighter when printing stacktrace in GUI

### LINKED ISSUE #3700: Error not shown proprly if error contains  `<` in text
## 🐛 Bug
Traceback from napari Traceback view:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/Projekty/PartSeg/package/PartSeg/_roi_analysis/measurement_widget.py in replace_measurement_result(self=)
    248         self.measurements_storage.clear()
    249         self.previous_profile = ""
--> 250         self.append_measurement_result()
        self.append_measurement_result = >
    251 
    252     def refresh_view(self):

~/Projekty/PartSeg/package/PartSeg/plugins/napari_widgets/measurement_widget.py in append_measurement_result(self=)
     54         if self.mask_chose.value is not None:
     55             image.mask = self.mask_chose.value.data
---> 56         roi_info = ROIInfo(self.roi_chose.value)
        roi_info = undefined
        global ROIInfo = 
        self.roi_chose.value = 
     57         dial = ExecuteFunctionDialog(
     58             compute_class.calculate,

~/Projekty/PartSeg/package/PartSegCore/roi_info.py in __init__(self= instance, roi=, annotations={}, alternative=None)
     58             return
     59         max_val = np.max(roi)
---> 60         dtype = minimal_dtype(max_val)
        dtype = undefined
        global minimal_dtype = 
        max_val = 
     61         roi = roi.astype(dtype)
     62         self.roi = roi

~/Projekty/PartSeg/package/PartSegImage/image.py in minimal_dtype(val=)
     24     :rtype:
     25     """
---> 26     if val  250:
        val = 
     27         return np.uint8
     28     if val  2 ** 16 - 5:

TypeError: '
```

Traceback from terminal:

```
TypeError                                 Traceback (most recent call last)
~/Projekty/PartSeg/package/PartSeg/_roi_analysis/measurement_widget.py in replace_measurement_result(self=<PartSeg.plugins.napari_widgets.measurement_widget.NapariMeasurementWidget object>)
    248         self.measurements_storage.clear()
    249         self.previous_profile = ""
--> 250         self.append_measurement_result()
        self.append_measurement_result = <bound method NapariMeasurementWidget.append_measurement_result of <PartSeg.plugins.napari_widgets.measurement_widget.NapariMeasurementWidget object at 0x7fb595b12160>>
    251 
    252     def refresh_view(self):

~/Projekty/PartSeg/package/PartSeg/plugins/napari_widgets/measurement_widget.py in append_measurement_result(self=<PartSeg.plugins.napari_widgets.measurement_widget.NapariMeasurementWidget object>)
     54         if self.mask_chose.value is not None:
     55             image.mask = self.mask_chose.value.data
---> 56         roi_info = ROIInfo(self.roi_chose.value)
        roi_info = undefined
        global ROIInfo = <class 'PartSegCore.roi_info.ROIInfo'>
        self.roi_chose.value = <Labels layer 'ROI' at 0x7fb595b055b0>
     57         dial = ExecuteFunctionDialog(
     58             compute_class.calculate,

~/Projekty/PartSeg/package/PartSegCore/roi_info.py in __init__(self=<class 'PartSegCore.roi_info.ROIInfo'> instance, roi=<Labels layer 'ROI'>, annotations={}, alternative=None)
     58             return
     59         max_val = np.max(roi)
---> 60         dtype = minimal_dtype(max_val)
        dtype = undefined
        global minimal_dtype = <function minimal_dtype at 0x7fb5f44b6ee0>
        max_val = <Labels layer 'ROI' at 0x7fb595b055b0>
     61         roi = roi.astype(dtype)
     62         self.roi = roi

~/Projekty/PartSeg/package/PartSegImage/image.py in minimal_dtype(val=<Labels layer 'ROI'>)
     24     :rtype:
     25     """
---> 26     if val < 250:
        val = <Labels layer 'ROI' at 0x7fb595b055b0>
     27         return np.uint8
     28     if val < 2 ** 16 - 5:

TypeError: '<' not supported between instances of 'Labels' and 'int'
```
As You could see, when  `<` occurs these parts are not properly rendered

## To Reproduce

1. Raise exception with `<` in the description in napari viewer.
2.
3.

## Expected behavior

Same text in napari window and console.

## Environment
napari: 0.4.12
Platform: Linux-5.11.0-40-generic-x86_64-with-glibc2.2.5
System: Ubuntu 21.04
Python: 3.8.3 (default, Sep 3 2021, 01:19:46) [GCC 10.3.0]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.21.3
SciPy: 1.7.1
Dask: 2021.09.1
VisPy: 0.9.4

OpenGL:
- GL version: 4.6.0 NVIDIA 470.82.00
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 2560x1440, scale 1.0
- screen 2: resolution 1920x1080, scale 1.0
- screen 3: resolution 1920x1080, scale 1.0

Plugins:
- PartSeg Analysis project reader: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg Image reader: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg Image tiff writer: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg Mask project reader: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg Mask project writer: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg Masked Tiff reader: 0.13.13.dev37+g429df8fb.d20211123
- PartSeg widgets: 0.13.13.dev37+g429df8fb.d20211123
- console: 0.0.4
- svg: 0.1.5

Happen also on the last development commit

## Additional context

This may be connected with using Rich text viewer which treats `<` as begin of formatting tag 

It is hard to search for issues by `traceback` in the description.

### LINKED ISSUE #4360: Blue text in plugin tracebacks is very low contrast
## 🐛 Bug
The dark blue text in error tracebacks from plugin errors is very low contrast against a black background, and therefore really hard to read:

Screenshot 2022-04-06 at 12 51 54

## To Reproduce

1. Trigger an error in a plugin

## Expected behavior

High contrast text 😄 

## Environment

 - Please copy and paste the information at napari info option in help menubar here:

 - Any other relevant information:

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Instead of pass HTML to Text area use `QtGui.QSyntaxHighlighter` with `pygments` 

This fix #3700 by omit using HTML and fix #4360 by using `syntax_style` from theme. 

Zrzut ekranu z 2022-04-25 11-34-45

Zrzut ekranu z 2022-04-25 11-34-22

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/theme.py [hunk 1]
```diff
@@ -294,7 +294,7 @@ def rebuild_theme_settings():
                 'warning': 'rgb(153, 18, 31)',
                 'current': 'rgb(0, 122, 204)',
                 'syntax_style': 'native',
-                'console': 'rgb(0, 0, 0)',
+                'console': 'rgb(18, 18, 18)',
```
Comment (Czaki) [inline napari/utils/theme.py:297]: follow material design suggestions to not use the black background. For me looks better 
https://material.io/design/color/dark-theme.html#properties

File: setup.cfg [hunk 2]
```diff
@@ -63,6 +63,7 @@ install_requires =
     PyOpenGL>=3.1.0
     PyYAML>=5.1
     pydantic>=1.9.0
+    pygments
```
Comment (Czaki) [inline setup.cfg]: already included by napari console

File: setup.cfg [hunk 3]
```diff
@@ -63,11 +63,12 @@ install_requires =
...
     qtpy>=1.7.0
     scipy>=1.4.1 ; python_version < '3.9'
     scipy>=1.5.4 ; python_version >= '3.9'
     superqt>=0.2.5
-    tifffile>=2020.2.16
+    tifffile>=2020.2.16,<2022.04.22
```
Comment (Czaki) [inline setup.cfg]: ` tifffile 2022.4.22` breaks test because of breaking change

**Comment by Czaki** [comment]
Based on information from https://github.com/napari/napari/issues/4410 test will fail until the new imagio release.

File: napari/_qt/dialogs/qt_notification.py [hunk 4]
```diff
@@ -344,7 +348,13 @@ def show_tb(parent):
                 tbdialog.setLayout(QVBoxLayout())
 
                 text = QTextEdit()
-                text.setHtml(notification.as_html())
+                with warnings.catch_warnings():
```
Comment (jni) [inline napari/_qt/dialogs/qt_notification.py]: Could you please add a comment about which warning this is silencing? This will help future maintenance in updating API calls etc.

File: napari/_qt/dialogs/qt_notification.py [hunk 5]
```diff
@@ -25,8 +26,11 @@
 )
 from superqt import QElidingLabel, ensure_main_thread
 
+from ...settings import get_settings
 from ...utils.notifications import Notification, NotificationSeverity
+from ...utils.theme import get_theme
 from ...utils.translations import trans
+from ..pyhon_syntax_highlight import Pylighter
```
Comment (jni) [inline napari/_qt/dialogs/qt_notification.py]: Typo: rename to `python_syntax_highlight`?

File: napari/_qt/pyhon_syntax_highlight.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        _format.setForeground(_color)
+
+    if "bold" in style and "nobold" not in style:
+        _format.setFontWeight(QtGui.QFont.Bold)
+    if "italic" in style and "noitalic" not in style:
+        _format.setFontItalic(True)
+    if "underline" in style and "nounderline" not in style:
+        _format.setFontUnderline(True)
+
+    return _format
```
Comment (jni) [inline napari/_qt/pyhon_syntax_highlight.py]: Could you add more comments about what is happening in this function? I find it hard to follow. Perhaps a docstring that explains the input will help. Also, it would be better if the variable name `_format` didn't shadow the function name (also `_format`).

File: napari/_qt/pyhon_syntax_highlight.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    def format(self, tokensource, outfile):
+        self.data = []
+
+        for token, value in tokensource:
+            self.data.extend(
+                [
+                    self._style[str(token)],
+                ]
+                * len(value)
+            )
```
Comment (jni) [inline napari/_qt/pyhon_syntax_highlight.py]: I find it confusing that `outfile` is not used at all, and instead an attribute is created (?) and then extended. Documentation would help here also.

File: napari/_qt/pyhon_syntax_highlight.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+from pygments import highlight
+from pygments.formatter import Formatter
+from pygments.lexers import get_lexer_by_name
+from pygments.styles import get_style_by_name
+from qtpy import QtGui
+
+# https://github.com/Vector35/snippets/blob/master/QCodeEditor.py
```
Comment (jni) [inline napari/_qt/pyhon_syntax_highlight.py]: Maybe add reference, "Using here under terms of MIT license"

File: napari/_qt/pyhon_syntax_highlight.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        self.formatter = QFormatter(theme)
+        self.lexer = get_lexer_by_name(lang)
+
+    def highlightBlock(self, text):
+        cb = self.currentBlock()
+        p = cb.position()
+        text = self.document().toPlainText() + '\n'
+        highlight(text, self.lexer, self.formatter)
+
+        # dirty, dirty hack
```
Comment (jni) [inline napari/_qt/pyhon_syntax_highlight.py]: This doesn't look good 😂 Could you elaborate on it?

**Review by jni — APPROVED** [review]
Thanks @Czaki! I've left some comments, all are about better documenting what the code is doing. Otherwise I agree with you that it would be really great to get this into 0.4.16, as it is a major usability improvement!

File: napari/_qt/pyhon_syntax_highlight.py (hunk 9 above)
Comment (Czaki) [reply napari/_qt/pyhon_syntax_highlight.py]: Could you check the elaboration?

File: napari/_qt/pyhon_syntax_highlight.py (hunk 6 above)
Comment (Czaki) [reply napari/_qt/pyhon_syntax_highlight.py]: write from scratch

File: napari/_qt/dialogs/qt_notification.py (hunk 5 above)
Comment (Czaki) [reply napari/_qt/dialogs/qt_notification.py]: ass one could pass language I cange it to `code_syntax_highlight`. Maybe it could be part of the public API for plugin creators? or migrated to superqt (after 0.4.16 release)?

File: napari/_qt/dialogs/qt_notification.py (hunk 4 above)
Comment (Czaki) [reply napari/_qt/dialogs/qt_notification.py]: Stupid copy and paste from another code part. After the check fixed to not use the warnings filter.

**Comment by Czaki** [comment]
> Thanks @Czaki! I've left some comments, all are about better documenting what the code is doing. Otherwise I agree with you that it would be really great to get this into 0.4.16, as it is a major usability improvement!

You force me to better understand this code (less inspiration, more documentation, and code digging) I hope that it improves code quality.

File: napari/_qt/dialogs/qt_notification.py [hunk 10]
```diff
@@ -344,7 +347,13 @@ def show_tb(parent):
                 tbdialog.setLayout(QVBoxLayout())
 
                 text = QTextEdit()
-                text.setHtml(notification.as_html())
+                theme = get_theme(
+                    get_settings().appearance.theme, as_dict=False
+                )
+                _highlight = Pylighter(  # noqa: F841
+                    text.document(), "python", theme.syntax_style
```
Comment (Czaki) [inline napari/_qt/dialogs/qt_notification.py:354]: I also check `pytb` language (`Python Traceback`) but it produces no formating

**Comment by Czaki** [comment]
Should HTML-related code be kept?

File: napari/_qt/dialogs/qt_notification.py (hunk 5 above)
Comment (jni) [reply napari/_qt/dialogs/qt_notification.py]: Yeah that would be good!

**Comment by jni** [comment]
@Czaki no, I don't see any reason why we should keep the html code, but I'm also ok with it happening in a subsequent PR. Thanks for all the improvements!

I will merge this before 0.4.16rc0. It's a wonderful usability improvement that's been bugging me for years! Just not enough because I typically work in IPython directly.

File: napari/_qt/dialogs/qt_notification.py (hunk 5 above)
Comment (Czaki) [reply napari/_qt/dialogs/qt_notification.py]: @tlambert03 What did you think about this. Did you prefer to have this as part of public API, or as part of superqt (there will be pygments dependency, maybe optional)?

**Comment by Czaki** [comment]
Changing font to monotype increases line length and I need to manually increase window size. 

This is not a blocking issue, but maybe someone with a better graphical interface projecting experience could take a look and decide if the default size of the window should be increased or the size of the font should be decreased?

File: napari/_qt/dialogs/qt_notification.py (hunk 5 above)
Comment (tlambert03) [reply napari/_qt/dialogs/qt_notification.py]: I like it,  happy for it to go into superqt.  fine if that happens later.  looks like pygments has no dependencies so probably fine just to add it directly

**Review by tlambert03 — APPROVED** [review]
nice improvement!
looks good, after changing the `pyhon` typo.

File: napari/_qt/dialogs/qt_notification.py (hunk 5 above)
Comment (Czaki) [reply napari/_qt/dialogs/qt_notification.py]: It may take time because next to code there should be documentation.

**Review by sofroniewn — APPROVED** [review]
LGTM! Thanks!!

**Comment by dstansby** [comment]
Amazing, thanks a lot for everyone who contributed to this PR 😄
